src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java

```diff
@@ -0,0 +1,140 @@
+    private final NetworkTableEntry velocity = table.getEntry("wheelvelocity");
+
+
+
+    private SwerveDriveKinematics kinematics;
+    private Odometry odometry;
+    
+    public boolean fieldCentric = false;
+
+    public SwerveDriveSubsystem (WheelSubsystem backRight, WheelSubsystem backLeft, WheelSubsystem frontRight, WheelSubsystem frontLeft, SwerveDriveKinematics kinematics, Odometry odometry) {
+        this.backRight = backRight;
+        this.backLeft = backLeft;
+        this.frontRight = frontRight;
+        this.frontLeft = frontLeft;
+
+        this.kinematics = kinematics;
+        this.odometry = odometry;
+
+        this.isFieldCentric.setBoolean(fieldCentric);
+    }
+
+
+    public void drive (double x, double y, double rot) {
+        
+
+        ChassisSpeeds speeds = new ChassisSpeeds(x, y, rot);
+        // ChassisSpeeds speeds = new ChassisSpeeds(0, 0, 0);
+
+        if (fieldCentric) {
+            speeds = ChassisSpeeds.fromFieldRelativeSpeeds(speeds, odometry.getRotation2d());
+        }
+        // Convert to module states
+        SwerveModuleState[] moduleStates = kinematics.toSwerveModuleStates(speeds);
+
+        SwerveModuleState frontLeftState = moduleStates[0];
+        SwerveModuleState frontRightState = moduleStates[1];
+        SwerveModuleState backLeftState = moduleStates[2];
+        SwerveModuleState backRightState = moduleStates[3];
+
+        velocity.setDouble(frontLeft.getSpeedMotor().getEncoder().getVelocity());
+
+        frontLeft.drive(frontLeftState);
+        frontRight.drive(frontRightState);
+        backLeft.drive(backLeftState);
+        backRight.drive(backRightState);
+
+    }
+
+    public void drive (SwerveModuleState[] moduleStates) {
+        SwerveModuleState frontLeftState = moduleStates[0];
+        SwerveModuleState frontRightState = moduleStates[1];
+        SwerveModuleState backLeftState = moduleStates[2];
+        SwerveModuleState backRightState = moduleStates[3];
+
+        velocity.setDouble(frontLeftState.speedMetersPerSecond);
+
+        frontLeft.drive(frontLeftState);
+        frontRight.drive(frontRightState);
+        backLeft.drive(backLeftState);
+        backRight.drive(backRightState);
+
+    }
+    
+
+    public void resetEncoders() {
+        frontLeft.getEncoder().reset();
+        frontLeft.getAngleMotor().getEncoder().setPosition(0);
+
+        frontRight.getEncoder().reset();
+        frontRight.getAngleMotor().getEncoder().setPosition(0);
+
+        backLeft.getEncoder().reset();
+        backLeft.getAngleMotor().getEncoder().setPosition(0);
+
+        backRight.getEncoder().reset();
+        backRight.getAngleMotor().getEncoder().setPosition(0);
+    }
+
+    @Override
+    public void periodic() {
+        odometry.update(getPositions());
+
+        odometryPose.setString(odometry.getPose().toString());
+
+        backLeftAngleEncoder.setDouble(backLeft.getEncoderPosition());
```

Comment:
These could also be part of the subsystem instead of the swerve drive system, no?

Reply:
If we put this in the wheelSubsystem, we can't keep each networkTableEntry separate.